### ISSUE TITLE
an attempt to impl zero-copy

### ISSUE THREAD

**Issue by zephyrchien** [zhboner/realm#31]
I think it's difficult to be compatible with tokio's `AsyncRead/AsyncWrite trait`, for that `AsyncRead/AsyncWrite` uses a `&[u8]` buffer however `splice` uses a pipe. 

So I've tried another approach. We can still use `TcpStream`, making use of its inner event loop. Instead of calling read/write, we can firstly get its inner fd via `as_raw_fd()` (`TcpStream` impls `AsRawFd`), then call `libc::splice` on the fd directly.

The problem is that `TcpStream::ready` does not clear the readiness on the fd, and there is no public fn to achieve this. I have to invoke `TcpStream::try_read/try_write` with an empty buffer to consume the read/write event.

Here's the demo
https://github.com/zephyrchien/realm/blob/bfed56a66d7c91126a41a05ad2bac7b92c728f61/src/relay.rs#L134-L215

**Comment by zhboner** [comment]
Zero copy would be a big improvement to the current implemention, but I don't think it is ready in tokio as mentioned in this issue https://github.com/tokio-rs/tokio/issues/2438

**Comment by zephyrchien** [comment]
Yup, zero-copy is of much significance, it would become a regret if Realm does not support this feature.

Tokio has its own difficulties, we could not expect Tokio to provide zero-copy APIs in the short term. But in our case, the problem could be simplified, making it possible to write a Tokio-based zero-copy API on our own.

Our async task is just to proxy 2 TCP connection, we never use buffers across tasks. **If we create a buffer(pipe) within the task, we could make sure that the pipe is always valid during the async operation.** When the task is finished or cancelled, the pipe will be dropped as normal.

Moreover, we could count the left bytes in the pipe, and limit the max read/write bytes when calling `splice`. This ensures that the pipe will never block, and we no longer need to poll the pipe, waiting for its readable/writeable event.
 
Given the inner fd of `tokio::TcpStream` has been registered to the event loop, the only thing we need to do is: 
1. wait the inner fd of `TcpStream` to become readable
2. splice fd1 -> pipe_write_half
3. wait the inner fd of another `TcpStream` to become writeable
4. splice pipe_read_half -> fd2

Actually I have did this a few days ago(0abbf2b [bin](https://github.com/zephyrchien/realm/releases/tag/custom-rc2)). I added these codes to `relay.rs` directly. I think the better approach is to encapsulate the pipe into a struct, and impl `Drop trait` for it, so that the pipe could be dropped automatically.
